From 2e0b8b44a9cc66b5873f17b4edbd478182382ede Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:37:01 -0400
Subject: [ethernet] i40e/i40evf: fix error checking path

Message-id: <1409658042-22689-61-git-send-email-sassmann@redhat.com>
Patchwork-id: 90254
O-Subject: [RHEL7.1 PATCH 060/281] i40e/i40evf: fix error checking path
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jean Sacren <sakiwit@gmail.com>

The commit 6494294f277fd ("i40e/i40evf: Use
dma_set_mask_and_coherent") uses dma_set_mask_and_coherent() to
replace dma_set_coherent_mask() for the benefit of return error.
The conversion brings some confusion in error checking as whether
against DMA_BIT_MASK(64) or DMA_BIT_MASK(32). For one, if error is
zero, the check will be against DMA_BIT_MASK(64) twice. Fix this
error checking by binding the check to the pertinent one.

Cc: Mitch Williams <mitch.a.williams@intel.com>
Signed-off-by: Jean Sacren <sakiwit@gmail.com>
Tested-by: Kavindya Deegala <kavindya.s.deegala@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit e3e3bfdd1d2ad4ac2020e681f3cb1e172f343810)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_main.c b/drivers/net/ethernet/intel/i40e/i40e_main.c
index 4241edc..efd82bf 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_main.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_main.c
@@ -8091,12 +8091,13 @@ static int i40e_probe(struct pci_dev *pdev, const struct pci_device_id *ent)
 
  /* set up for high or low dma */
  err = dma_set_mask_and_coherent(&pdev->dev, DMA_BIT_MASK(64));
- if (err)
-  err = dma_set_mask_and_coherent(&pdev->dev, DMA_BIT_MASK(32));
  if (err) {
-  dev_err(&pdev->dev,
-   "DMA configuration failed: 0x%x\n", err);
-  goto err_dma;
+  err = dma_set_mask_and_coherent(&pdev->dev, DMA_BIT_MASK(32));
+  if (err) {
+   dev_err(&pdev->dev,
+    "DMA configuration failed: 0x%x\n", err);
+   goto err_dma;
+  }
  }
 
  /* set up pci connections */
-- 
1.7.1